DFR-3063 Upload warnings on upload document

so99y opened the pull request:
### Jira link (if applicable)
https://tools.hmcts.net/jira/browse/DFR-3063


### Change description ###


### Checklist

<!-- Check each box by removing the space and adding an x, e.g. [x] -->

- [X] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [X] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

jthmcts commented on src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/util/TestResource.java:
```diff
@@ -38,4 +46,43 @@ public static Document testDocument() {
             .binaryUrl(BINARY_URL)
             .build();
     }
+
```
I like the idea but this isn't the right place for the code. Compare to this class which is achieving a similar goal

https://github.com/hmcts/finrem-case-orchestration-service/blob/master/src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/FinremCallbackRequestFactory.java

so99y replied:
I am gonna change the caller to use `FinremCallbackRequestFactory` class.

jthmcts commented on src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/DocumentUploadServiceV2.java:
```diff
@@ -0,0 +1,40 @@
+import static org.apache.commons.lang3.ObjectUtils.isEmpty;
+
+@Service
+public class DocumentUploadServiceV2 {
```
I don't think this name really really tells you what the purpose of the class is. Why would I use this V2 rather than the original? Is the V1 deprecated?

so99y replied:
Since the existing DocumentUploadService is an abstract class with a generic type parameter, it doesn't apply to my logic at all.

ptrelease replied:
Perhaps a comment at the top of v2 to explain the difference with V1?

jthmcts replied:
How about `NewUploadedDocumentsService`?

@so99y If you want to stick with the name and javadoc then ok - but the javadoc should at least make use of `@link` for class references.

so99y replied:
Sounds good. I will rename the service.

so99y commented on src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/DocumentUploadServiceV2Test.java:
```diff
@@ -0,0 +1,185 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@ExtendWith(SpringExtension.class)
+class DocumentUploadServiceV2Test extends BaseServiceTest {
```
Good sport. Updated acccordingly.

ptrelease commented on src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/documentchecker/contentchecker/ApplicantNameDocumentContentChecker.java:
```diff
@@ -0,0 +1,50 @@
+        ContactDetailsWrapper contactDetails = caseData.getContactDetailsWrapper();
+        return ofNullable(contactDetails.getApplicantFmName()).orElse(StringUtils.EMPTY) + SPACE
+            + ofNullable(contactDetails.getApplicantLname()).orElse(StringUtils.EMPTY);
+    }
```
Curious on why the return doesn’t format an interpolated string here. 
This works fine, just looks unusual. So far I’ve only seen the SPACE constant used here too.

so99y replied:
Yeah. I reviewed it and I made some changes on this bit. please take a look. (f52bf7fdf1afca5492988fbbe305778a6a382e46)

ptrelease commented:
Nothing much to add, looks good.  The uploadDocuments and uploadGeneralDocument field names for consented and contested aren’t informative - but I guess changing would mean a migration task? 
There is one little typo in the mock called InptuCaseDocument. But it’s referenced everywhere correctly, so not an issue.

so99y commented:
> @so99y Nothing much to add, looks good. The uploadDocuments and uploadGeneralDocument field names for consented and contested aren’t informative - but I guess changing would mean a migration task? There is one little typo in the mock called InptuCaseDocument. But it’s referenced everywhere correctly, so not an issue.

- It requires data migration if we want to change the field names. 
- I changed the typo bit.
- I updated the Document content checkers for the SPACE stuff and changed the method to return a formatted string instead.
- Will add a comment on DocumentServiceV2 to explain the difference
- Moving the TestResource methods to `FinremCallbackRequestFactory`

so99y commented:
@jthmcts I have added `FinremCaseDetailsBuilderFactory` and included some null-safe checkings on document content checkers.

al-hmcts commented:
Looks really good Ashely! I just have one question, and I may have the wrong end of the stick for the implementation. 

I can see you have used an interface to for the document checkers that defines the method which are called by the dependent service, however, not all of the classes that implement the interface utilise the method param `beforecaseDetails` of the interface defined `getWarnings(...)` method (namely DocxDocumentChecker.java and PdfDocumentChecker.java). 

Making the concrete class ignore unused arguments violates the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle), so could I suggest that we could implement a different pattern here to avoid this? We could potentially use the [Command Pattern](https://www.baeldung.com/java-command-pattern) here instead. 

Again, I my understanding of your implementation may be wrong, happy to have a chat 😄

ptrelease commented on src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/documentchecker/DuplicateFilenameDocumentChecker.java:
```diff
@@ -21,22 +32,94 @@ public boolean canCheck(CaseDocument caseDocument) {
+        processHasCaseDocument(caseData, allDocuments);
+        return allDocuments.stream().filter(Objects::nonNull).toList();
+    }
+
```
Can ln 53 be removed?
```
 if (hcd != null) {
```
Line 43 is doing the null check already as far as I can tell:

```
if (item instanceof HasCaseDocument hasCaseDocument) {
```

ptrelease replied:
Also, just to check, will the logs on line 46 and 78 be useful?  As in can we tie them back to a case if we were looking in Insights?

so99y replied:
Good spot. Yes, NPE has been handled and the log does not make any sense. I've just removed the check null logic and the log.

ptrelease commented on src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/documentchecker/DuplicateFilenameDocumentChecker.java:
```diff
@@ -0,0 +1,125 @@
+            }
+        }
+    }
+
```
This is complex.  The method name is fine, but some JavaDoc for the method would be great.  Something that explains what the method does, to save the next dev from having to work it out :-D

so99y replied:
Java doc added.

ptrelease commented on src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/documentchecker/DuplicateFilenameDocumentCheckerTest.java:
```diff
@@ -0,0 +1,972 @@
+                .consentVariationOrderDocument(DUPLICATED_CASE_DOCUMENT)
+                .build())
+        );
+    }
```
Are these all used?

ptrelease replied:
Ah, just clocked that it's a parameterized test.  So guess they are all used?

so99y replied:
Basically the arguments are setting a CaseDocument with the target filename to all possible locations in `FinremCaseData` including the nested fields and the element of a List. 

Therefore the same testing result will be expected in the test:
`void testGetWarningsOnExistingCase(FinremCaseData caseData)`

ptrelease commented on src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/documentchecker/DuplicateFilenameDocumentCheckerTest.java:
```diff
@@ -0,0 +1,972 @@
+        assertThat(warnings).isEmpty();
+    }
+
+    @Test
```
testGetWarningsOnUploadingDuplicatedFilesAtTheSameTimeContented. Probably mean Contested (not contented)

so99y replied:
sorry for the typo

ptrelease commented on src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/documentchecker/DuplicateFilenameDocumentChecker.java:
```diff
@@ -0,0 +1,125 @@
+
+        return Collections.emptyList();
+    }
+}
```
I notice the getWarnings here checks beforeCaseDetails and newCasedata.  I _think_ the unit tests just test newCaseData?  

The unit tests implicitly test these private functions, so sorry if I missed it.

so99y replied:
checking on `newCaseData` only happens when I check the duplication in new files which is covered in - 
`testGetWarningsOnUploadingDuplicatedFilesAtTheSameTimeConsented`
`shouldNotGetWarningsOnUploadingDuplicatedFilesAtTheSameTimeConsented`
`testGetWarningsOnUploadingDuplicatedFilesAtTheSameTimeContested`
`shouldNotGetWarningsOnUploadingDuplicatedFilesAtTheSameTimeContested`

so99y commented:
> Looks really good Ashely! I just have one question, and I may have the wrong end of the stick for the implementation.
> 
> I can see you have used an interface for the document checkers that defines the methods that are called by the dependent service, however, not all of the classes that implement the interface utilise the method param `beforecaseDetails` of the interface defined method `getWarnings(...)` (namely DocxDocumentChecker.java and PdfDocumentChecker.java).
> 
> Making the concrete class ignore unused arguments violates the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle), so could I suggest that we could implement a different pattern here to avoid this? We could potentially use the [Command Pattern](https://www.baeldung.com/java-command-pattern) here instead.
> 
> Again, my understanding of your implementation may be wrong, happy to have a chat 😄

It's because I am lazy. `beforecaseDetails` was just introduced to the interface and I think I should write it better. I will update the parameters accordingly.

so99y commented:
@al-hmcts I modified the parameters using the Builder pattern ([b84d6b5](https://github.com/hmcts/finrem-case-orchestration-service/pull/1731/commits/b84d6b5a158e1db5bf5d35b5ded5145153eae39a)) Please take a look. Thanks

al-hmcts reviewed (approved):
LGTM, happy with the builder pattern!

so99y commented on src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/UploadDocumentConsentedAboutToSubmitHandlerTest.java:
```diff
@@ -0,0 +1,155 @@
+            assertThat(logs.getInfos()).isEmpty();
+        }
+    }
+
```
Sounds good.

ptrelease commented on src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/documentchecker/DocumentCheckerService.java:
```diff
@@ -0,0 +1,56 @@
+                    .build()));
+            } catch (DocumentContentCheckerException e) {
+                log.error(format("Unexpected error when getting warnings from %s", dc.getClass().getName()), e);
+            }
```
Be great to have the case ID in this?

so99y replied:
Sounds great. https://github.com/hmcts/finrem-case-orchestration-service/pull/1731/commits/534a29139f2dda03dc1c5e6e34e1b1641bc75f0b

ptrelease commented on src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/documentchecker/DuplicateFilenameDocumentChecker.java:
```diff
@@ -0,0 +1,129 @@
+        List<DocumentFileNameProvider> allDocuments = collectCaseDocumentsFromFinremCaseData(context.getBeforeCaseDetails().getData());
+
+        log.info("Iterating all CaseDocuments with interface HasCaseDocument.");
+
```
Is this log useful for support?

so99y replied:
removed
